lib/srv/alpnproxy/conn.go

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2020-2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alpnproxy
+
+import (
+	"io"
+	"net"
+	"time"
+)
+
+// newBufferedConn creates new instance of bufferedConn.
+func newBufferedConn(conn net.Conn, header io.Reader) *bufferedConn {
+	return &bufferedConn{
+		Conn: conn,
+		r:    io.MultiReader(header, conn),
+	}
+}
+
+// bufferedConn allows injecting additional reader that will be drained during Read call reading from net.Conn.
+// Is used when part of the data on a connection has already been read.
+type bufferedConn struct {
```

Comment:
We have `lib/multiplexer/wrappers.go:Conn` that I think does the same. Have you tried using that one?

Reply:
I have looked at `lib/multiplexer/wrappers.go:Conn` but the usage seems to be a bit different. The wrappers.Conn allows to peek into the connection in order to read header bytes but in my case i want to append "header" bytes into connection so next Read connection call will drain the header reader before reading bytes from net.Conn.